R/tags.R

```diff
@@ -0,0 +1,125 @@
+
+#' Get a simple data.frame listing all Tags on server
+#' 
+#' @param src The source object
+#' @param use_cache use the tag list previously queried 
+#' from the connect server. src$tags object is null then
+#' this parameter will be ignored
+#' 
+#' @export
+get_tags <- function(src, use_cache = FALSE){
+  validate_R6_class("Connect", src)
+  
+  res <- src$get_tags(use_cache = use_cache)
+  
+  return(res)
+}
+
+#' Get a tibble with more information for each tag
+#' 
+#' @param src The source object
+#' 
+#' @export
+get_tag_info <- function(src){
+  validate_R6_class("Connect", src)
+  
+  res <- src$get_tag_tree()
+  
+  tag_tbl <- parse_tags_tbl(res)
+  
+  return(tag_tbl)
+}
+
+#' Print a visualization of the tag tree
+#' 
+#' @param src The source object
+#' @param top_tag The value to be printed at the top of the 
+#' tag tree. This value can not have any "/" characters in it. 
+#' 
+#' @export
+get_tag_tree <- function(src, top_tag = "RStudio Connect Tags"){
+  validate_R6_class("Connect", src)
+  
+  res <- src$get_tag_tree()
+  
+  tag_tree(res, top_tag = top_tag)
```

Comment:
So this currently returns `NULL`, which seems a bit hard to "use." What are your thoughts on either:
- inventing an S3 class with a `print` method that gives you this pretty "tree" output
- making this more of a `print` or `show` method name, since you can't actually _use_ it for anything but illustration

If you look carefully, this is the pattern followed in `fs`. I'm also happy to accept this PR as-is and work on that change in a follow-up, if that's preferable, though! It's a bit of rearchitecting to be done 😄 This is a really cool picture of the nested tag list!

```
# fs example - sorry. Lazy and not a reprex
> hm <- fs::dir_tree("1558583138/")
1558583138/
> hm
1558583138/get_files
1558583138/get_files/rstudio-connect.gcfg
1558583138/get_files/rstudio-connect.gcfg/Failed
1558583138/get_files/rstudio-connect.gcfg/Failed/i-06fa8f5a99359338f
1558583138/get_files/rstudio-connect.gcfg/Failed/i-070464f64bc2fe5b2
1558583138/get_files/rstudio-connect.gcfg/Failed/i-08bffa9cb088d3093
1558583138/get_files/rstudio-connect.gcfg/Failed/i-09cf956e63c58753a
1558583138/get_files/rstudio-connect.gcfg/Failed/i-09d05131561dc044d
1558583138/get_files/rstudio-connect.gcfg/Failed/i-0b6a4a8df217ecb3d
1558583138/get_files/rstudio-connect.gcfg/Failed/i-0bbdd9a4ef5108a20
1558583138/get_files/rstudio-connect.gcfg/Failed/i-0c6933ba1d3d27a31
1558583138/get_files/rstudio-connect.gcfg/Failed/i-0c9dbdcc13dd5910b
1558583138/get_files/rstudio-connect.gcfg/Failed/i-0d3b4c10e58a50c55
1558583138/get_files/rstudio-connect.gcfg/Failed/i-0d3e90864c93d7992
1558583138/get_files/rstudio-connect.gcfg/Failed/i-0d4aa44abb1d5b83b
1558583138/get_files/rstudio-connect.gcfg/Failed/i-0d5160ac485a3fa60
1558583138/get_files/rstudio-connect.gcfg/Failed/i-0d616c0dc35382389
1558583138/get_files/rstudio-connect.gcfg/Success
1558583138/get_files/rstudio-connect.gcfg/Success/i-08497d121cbdd253a
1558583138/get_files/rstudio-connect.gcfg/Success/i-0a2030206cd446d46
> class(hm)
[1] "fs_path"   "character"

# connectapi
> obj <- get_tag_tree(client)
RStudio Connect Tags
├── Technology
│   ├── Deep Learning
│   │   ├── keras
│   │   ├── tensorflow
│   │   └── Pre-trained Models
│   ├── Output Types
│   │   ├── Excel
│   │   ├── PDF
│   │   └── PowerPoint
│   ├── Email
│   ├── BackEnds
│   │   ├── Databases
│   │   └── Spark
│   └── Dashboards
│       ├── flexdashboard
│       ├── shinydashboard
│       ├── D3
│       └── Maps
├── Projects and Presentations
│   ├── Calendars
│   ├── QuickStart
│   │   └── Stock Portfolio
│   ├── Production Webinar
│   │   ├── Production
│   │   └── Prototype
│   ├── Pins Demo: Bikes
│   │   ├── Pins
│   │   └── Other Assets
│   └── Pro Admin Training
└── Solutions
    ├── Access to Care
    ├── Python
    ├── Immunotherapy
    ├── Customer Churn
    └── Model Management
> obj
NULL
```

Reply:
Related:
```
> fs:::print.fs_path
function (x, ..., max = getOption("max.print")) 
{
    x <- x[seq_len(min(length(x), max))]
    if (length(x) == 0) {
        print(character(0))
    }
    else {
        cat(multicol(colourise_fs_path(x, ...)), sep = "")
    }
    invisible(x)
}
<bytecode: 0x7f7fe143cdc8>
<environment: namespace:fs>
```

Reply:
What if we simply add `invisible(parse_tags_tbl(res))` to the end of the function which would return:

```
client <- connectapi::connect()
 
tmp <- get_tag_tree(client)
#> RStudio Connect Tags
#> +-- BLS
#> |   \-- SRA
#> \-- OOW
#>     \-- H&H
 
tmp
#> # A tibble: 4 x 6
#>      id name  created_time         updated_time         version parent_id
#>   <int> <chr> <chr>                <chr>                  <int>     <int>
#> 1     1 BLS   2019-10-24T18:48:01Z 2019-10-24T18:48:01Z       0        NA
#> 2     4 SRA   2019-10-24T18:48:21Z 2019-10-24T18:48:21Z       0         1
#> 3     2 OOW   2019-10-24T18:48:12Z 2019-10-24T18:48:12Z       0        NA
#> 4     3 H&H   2019-10-24T18:48:16Z 2019-10-24T18:48:16Z       0         2
```

Reply:
Or we could return the unparsed list too with `invisible(res)` if you think people would want to see it in the nested list format